File: Beecrowd_Parte5/Time_De_Duendes_3176.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    char nome[21];
    int idade;
} Duende;

// Função de comparação para qsort
int comparar(const void *a, const void *b) {
    Duende *duendeA = (Duende *)a;
    Duende *duendeB = (Duende *)b;
    if (duendeB->idade != duendeA->idade) {
        return duendeB->idade - duendeA->idade;  // Decrescente por idade
    } else {
        return strcmp(duendeA->nome, duendeB->nome);  // Crescente por nome
    }
}

int main() {
    int N;
    scanf("%d", &N);
    
    Duende duendes[N];
    
    // Lendo os nomes e idades dos duendes
    for (int i = 0; i < N; i++) {
        scanf("%s %d", duendes[i].nome, &duendes[i].idade);
    }
    
    // Ordenando a lista de duendes
    qsort(duendes, N, sizeof(Duende), comparar);
    
    // Número de times
    int T = N / 3;
    
    // Formando e imprimindo os times
    for (int i = 0; i < T; i++) {
        printf("Time %d\n", i + 1);
        printf("%s %d\n", duendes[i].nome, duendes[i].idade);
        printf("%s %d\n", duendes[i + T].nome, duendes[i + T].idade);
        printf("%s %d\n", duendes[i + 2 * T].nome, duendes[i + 2 * T].idade);
        printf("\n");
    }
    
    return 0;
}

File: Beecrowd_Parte5/Time_De_Duendes_3176.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Scanner;

class Duende {
    String nome;
    int idade;

    public Duende(String nome, int idade) {
        this.nome = nome;
        this.idade = idade;
    }
}

public class Time_De_Duendes_3176 {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Lendo o número de duendes
        int N = scanner.nextInt();
        scanner.nextLine(); // Consumir a nova linha

        List<Duende> duendes = new ArrayList<>();

        // Lendo os nomes e idades dos duendes
        for (int i = 0; i < N; i++) {
            String nome = scanner.next();
            int idade = scanner.nextInt();
            duendes.add(new Duende(nome, idade));
        }

        // Ordenar a lista de duendes por idade decrescente e, em caso de empate, por nome crescente
        Collections.sort(duendes, new Comparator<Duende>() {
            @Override
            public int compare(Duende d1, Duende d2) {
                if (d2.idade != d1.idade) {
                    return Integer.compare(d2.idade, d1.idade);
                } else {
                    return d1.nome.compareTo(d2.nome);
                }
            }
        });

        // Número de times
        int T = N / 3;

        // Formar e imprimir os times
        for (int i = 0; i < T; i++) {
            System.out.println("Time " + (i + 1));
            System.out.println(duendes.get(i).nome + " " + duendes.get(i).idade);
            System.out.println(duendes.get(i + T).nome + " " + duendes.get(i + T).idade);
            System.out.println(duendes.get(i + 2 * T).nome + " " + duendes.get(i + 2 * T).idade);
            System.out.println();
        }

        scanner.close();
    }
}
